include/oneapi/dpl/pstl/hetero/numeric_impl_hetero.h

```diff
@@ -152,18 +152,77 @@ __pattern_transform_scan_base_impl(_ExecutionPolicy&& __exec, _Iterator1 __first
     }
 }
 
+// TODO In C++20 we may try to use std::equality_comparable
+template <typename _Iterator1, typename _Iterator2, typename = void>
+struct is_equality_comparable : std::false_type
+{
+};
+
+// Two sycl_iterator's (with different access modes)
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+struct is_equality_comparable<sycl_iterator<_Mode1, _T, _Allocator>, sycl_iterator<_Mode2, _T, _Allocator>>
+    : std::true_type
+{
+};
+
+// All with implemented operator ==
 template <typename _Iterator1, typename _Iterator2>
-constexpr bool
+struct is_equality_comparable<_Iterator1, _Iterator2,
+                              std::void_t<decltype(::std::decay_t<_Iterator1>() == ::std::decay_t<_Iterator2>())>>
+    : std::true_type
+{
+};
+
+// Check the ability to compare two sycl_iterator's
+/*
+ * @return bool - true, if these iterators may be the same;
+ *         false - we know these iterators are different
+ */
+template <sycl::access::mode _Mode1, sycl::access::mode _Mode2, typename _T, typename _Allocator>
+bool
+__check_if_iterator_equality_is_possible(sycl_iterator<_Mode1, _T, _Allocator> __it1,
+                                         sycl_iterator<_Mode2, _T, _Allocator> __it2)
+{
+    const auto buf1 = __it1.get_buffer();
+    const auto buf2 = __it2.get_buffer();
+
+    // We are unable to compare two sycl_iterator's from different buffers
+    if (buf1 != buf2)
+        return false;
+
+    // We are unable to compare two sycl_iterator's if one of them is sub_buffer
+    if (buf1.is_sub_buffer() || buf2.is_sub_buffer())
```

Comment:
SYCL specific code os not possible here, on `dpl/pstl/hetero` level, just on `dpl/pstl/hetero/dpcpp` level.

Reply:
And actually, why cannot we compare two iterators if one of them is over a sub-buffer?

Reply:
Fixed: if one is sub_buffer, continue like in case when iterators are equal.